### PR TITLE
KTOR-7489 Multi-platform support + kotlinx-rpc

### PR DESCRIPTION
**Pull request by bjhham**
Got a few changes in here...

1. Dependency resolution and compilation of all plugins are now handled through gradle during the build phase.  This allows for highlighting and auto-complete to work in all the plugin source directories, using only the yaml manifests.
2. Each plugin can have multiple modules defined, based on some pre-defined modules: server (default), core, and client.  These are resolved as separate gradle modules with independent dependencies and sources.
3. Modules can be referenced from the plugins for dependencies, gradle plugins, and source files.  This controls where the generator will send these items.
4. Deleted some 1.* files and removed support for the old JSON formats.
5. Moved around tests and related artifacts because of more logic happening during the build phase.
6. Added ability to disable build systems for plugins when unsupported (i.e., maven, amper)
7. Moved version variables from yaml to plugins/gradle.properties, so it's easier for dependabot etc.  Also added version references from manifest for gradle plugins
8. Added kotlinx-rpc plugin 🎉

### CODE REVIEW

File: shared/io/ktor/plugins/registry/Properties.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import kotlin.io.path.inputStream
+
+fun Path.readVersionProperties() = resolve("gradle.properties").let { versionsFile ->
+    try {
+        Properties().apply {
+            load(versionsFile.inputStream())
+        }.entries.associate { (key, value) ->
+            key.toString() to value.toString()
+        }
+    } catch (_: Exception) {
```
Comment (e5l) [inline shared/io/ktor/plugins/registry/Properties.kt:20]: Throwable

File: shared/io/ktor/plugins/registry/PluginReference.kt [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    override fun serialize(encoder: kotlinx.serialization.encoding.Encoder, value: ArtifactReference) {
+        encoder.encodeString(value.toString())
+    }
+
+    override fun deserialize(decoder: kotlinx.serialization.encoding.Decoder): ArtifactReference =
+        ArtifactReference.parse(decoder.decodeString())
+}
+
+// TODO maintain variables
```
Comment (e5l) [inline shared/io/ktor/plugins/registry/PluginReference.kt:212]: please log an issue for TODO

File: shared/io/ktor/plugins/registry/PluginReference.kt [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+fun prefixVersionToMavenRange(text: String): String {
+    require(text.endsWith(".+")) { "Expected \"$text\" to end with .+" }
+    val prefix = text.substringBeforeLast(".+")
+    val nextPrefix = prefix.split('.').mapIndexed { index, part ->
+        if (index < prefix.count { it == '.' }) part else (part.toInt() + 1).toString()
+    }.joinToString(".")
+    return "[$prefix,$nextPrefix)"
+}
+
+class ArtifactReferenceSerializer: kotlinx.serialization.KSerializer<ArtifactReference> {
```
Comment (e5l) [inline shared/io/ktor/plugins/registry/PluginReference.kt:197]: could you tell me why do we need package name here?

File: plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+name: kotlinx.rpc
+description: Adds remote procedure call (RPC) routing
+vcsLink: https://github.com/Kotlin/kotlinx-rpc
+license: Apache 2.0
+category: Frameworks
+maven:
+  disabled: true
+# TODO waiting for fixes to enable amper support
```
Comment (e5l) [inline plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml:8]: please log issue for this as well

File: plugins/client/io.github.flaxoos/ktor-client-circuit-breaker/3.0/install.kt [hunk 5]
```diff
@@ -1,17 +1,8 @@
-/*
- * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
- */
-
 import io.github.flaxoos.ktor.client.plugins.circuitbreaker.CircuitBreakerName.Companion.toCircuitBreakerName
```
Comment (e5l) [inline plugins/client/io.github.flaxoos/ktor-client-circuit-breaker/3.0/install.kt:1]: Please configure copyright in IntelliJ and run action to generate copyright back for the repo

File: buildSrc/src/test/kotlin/io/ktor/plugins/registry/FilesTest.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        }.flatMap { versionsFile ->
+            versionsFile.readYamlMap()?.let { yamlMap ->
+                yamlMap.entries.asSequence()
+                    .filter { (key) -> key.content.matches(Regex("[\\w-]+")) }
+                    .map { (key, value) -> "${key.content}=${value.yamlScalar.content}" }
+            }.orEmpty()
+        }.distinctBy {
+            it.substringBefore('=')
+        }.forEach { entry ->
+            println(entry)
```
Comment (e5l) [inline buildSrc/src/test/kotlin/io/ktor/plugins/registry/FilesTest.kt:34]: println

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginResolution.kt [hunk 7]
```diff
@@ -5,80 +5,48 @@
...
-    val artifactsByRelease =
-        configurations.mapValues { (_, artifacts) ->
-            artifacts.associate { resolvedArtifact ->
-                Pair(
-                    resolvedArtifact.referenceString(),
-                    resolvedArtifact.file.path
-                )
+    for (pluginConfig in pluginConfigs) {
+        val classPathFile = classpathsDir.resolve("${pluginConfig.id}.${pluginConfig.release}.yaml")
+        val alreadyResolved: Set<String> = if (classPathFile.exists()) {
```
Comment (e5l) [inline buildSrc/src/main/kotlin/io/ktor/plugins/registry/PluginResolution.kt:30]: I would extract such blocks in a separate functions to improve readablility

**Review by e5l — APPROVED** [review]
lgtm, left minor comments
Please consider extracting plugin fixed and new plugins to separate PRs next time

File: plugins/client/io.github.flaxoos/ktor-client-circuit-breaker/3.0/install.kt (hunk 5 above)
Comment (bjhham) [reply plugins/client/io.github.flaxoos/ktor-client-circuit-breaker/3.0/install.kt:1]: I'm not sure we want to include the copyright for the plugin sample code, do we?  I think the code within each plugin ought to be owned by the contributor and licensed based on the manifest.

File: plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml (hunk 4 above)
Comment (bjhham) [reply plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml:8]: There is one now, I'll add it to the TODO

File: shared/io/ktor/plugins/registry/Properties.kt (hunk 1 above)
Comment (bjhham) [reply shared/io/ktor/plugins/registry/Properties.kt:20]: I think it'd be better to crash here if it's a non-exception throwable.

File: shared/io/ktor/plugins/registry/PluginReference.kt (hunk 2 above)
Comment (bjhham) [reply shared/io/ktor/plugins/registry/PluginReference.kt:212]: Oh, I had done it, just forgot to remove the TODO.

File: shared/io/ktor/plugins/registry/PluginReference.kt (hunk 3 above)
Comment (bjhham) [reply shared/io/ktor/plugins/registry/PluginReference.kt:197]: I'll drop them here, Intellij added them as I was moving stuff around.

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/Files.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import kotlin.io.path.ExperimentalPathApi
+import kotlin.io.path.createDirectory
+import kotlin.io.path.deleteRecursively
+import kotlin.io.path.exists
+import kotlin.io.path.inputStream
+import kotlin.io.path.isDirectory
+import kotlin.io.path.listDirectoryEntries
+
+// Get all directories using patterns like "plugins/client/*/*"
+fun folders(pattern: String): List<Path> {
```
Comment (Mr3zee) [inline buildSrc/src/main/kotlin/io/ktor/plugins/registry/Files.kt:22]: There is a thing like Files.listDirs with glob pattern, can we use it here?

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/Files.kt [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                }
+            }
+            else -> folders.mapNotNull { folder -> folder.resolve(Paths.get(next)).takeIf { it.exists() } }
+        }
+    }
+    return folders.filter { it.isDirectory() }
+}
+
+@OptIn(ExperimentalPathApi::class)
+fun Path.clear(): Path {
```
Comment (Mr3zee) [inline buildSrc/src/main/kotlin/io/ktor/plugins/registry/Files.kt:44]: naming does not imply creating new dir

File: buildSrc/src/main/kotlin/io/ktor/plugins/registry/Strings.kt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.plugins.registry
+
+fun String.stripSpecialChars() =
```
Comment (Mr3zee) [inline buildSrc/src/main/kotlin/io/ktor/plugins/registry/Strings.kt:7]: What is special?

File: plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  plugins:
+    - id: org.jetbrains.kotlin.plugin.serialization
+      version: LAST_KOTLIN_VERSION
+      module: core
+    - id: org.jetbrains.kotlinx.rpc.plugin
+      version: $kotlinx-rpc
+      module: core
+    - id: org.jetbrains.kotlin.plugin.serialization
+      version: LAST_KOTLIN_VERSION
+      module: client
```
Comment (Mr3zee) [inline plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml:22]: No need to that I think, only for the module with code generation

File: plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    - id: org.jetbrains.kotlinx.rpc.plugin
+      version: $kotlinx-rpc
+      module: client
+    - id: org.jetbrains.kotlinx.rpc.plugin
+      version: $kotlinx-rpc
+      module: server
```
Comment (Mr3zee) [inline plugins/server/org.jetbrains/kotlinx-rpc/3.0/manifest.ktor.yaml:28]: Same here

**Review by Mr3zee — APPROVED** [review]
Amazing job! One thing is that my review is a bit shallow, as the PR is huge. But DM helped a lot. Below are some minor comments

**Pull request merged**